GowerMetric.py

```diff
@@ -0,0 +1,379 @@
+import os.path
+import timeit
+from pathlib import Path
+from typing import List, Optional, Union
+
+import numpy as np
+import sklearn.utils.validation
+from scipy.spatial.distance import pdist
+from scipy.cluster.hierarchy import (
+    linkage,
+    cophenet,
+    fcluster,
+)
+from scipy.optimize import minimize
+from sklearn.metrics import silhouette_score
+from sklearn.preprocessing import OrdinalEncoder, StandardScaler
+
+from utils import DataType
+
+
+class GowerMetric:
+    def __init__(
+        self,
+        dtypes: np.array,
+        ratio_scale_normalization: str = "range",
+        weights: Optional[Union[list, str]] = None,
+        precomputed_weights_file: Optional[str] = None,
+    ):
+        assert (
+            weights is None
+            or weights == "precomputed"
+            or weights == "cpcc"
+            or type(weights) == np.ndarray
+            or type(weights) == list
+        )
+        assert (
+            ratio_scale_normalization == "range"
+            or ratio_scale_normalization == "iqr"
+            or ratio_scale_normalization == "kde"
+        )
+
+        self.dtypes = dtypes  # initialize with np.array of column data types
+        self.weights = weights
+        self.precomputed_weights_file = precomputed_weights_file
+        self.ranges_: np.ndarray  # values of ranges in .ratio_scale() (iqr or traditional range)
+        self.h_: np.ndarray  # h values in .ratio_scale()
+        self.n_features_in_: int
+        self.ratio_scale_normalization: str = ratio_scale_normalization
+
+        # Bit masks for certain column types
+        self.ratio_scale_idx = self.dtypes == DataType.RATIO_SCALE
+        self.cat_nom_idx = (self.dtypes == DataType.CATEGORICAL_NOMINAL) | (
+            self.dtypes == DataType.BINARY_SYMMETRIC
+        )
+        self.bin_asym_idx = self.dtypes == DataType.BINARY_ASYMMETRIC
+
+        # Sums of columns of given type
+        self.cat_nom_num = sum(self.cat_nom_idx)
+        self.ratio_scale_num = sum(self.ratio_scale_idx)
+        self.bin_asym_num = sum(self.bin_asym_idx)
+
+    def fit(self, X):
+        assert X.shape[1] == len(self.dtypes)
+
+        self.ranges_ = np.empty(self.ratio_scale_num)
```

Comment:
You do not need to create this attribute here, you overwrite it a bit lower anyway